### PR TITLE
feat: add apply_token_bitmask operator

### PR DESCRIPTION
**Pull request by CarterDuan**
This PR introduces the apply_token_bitmask operator.
When speculative decoding is enabled in SGLang and a user request requires a JSON Schema output format, the xgrammar_backend will be triggered. It contains the operator apply_token_bitmask_inplace_triton, which cannot run on NPU. The alternative CPU fallback solution suffers from poor performance, so an AscendC-based apply_token_bitmask operator is implemented.
Integration into SGLang has been completed, along with single-concurrency validation and performance benchmarking.
In single-concurrency scenarios, the average latency of the CPU fallback implementation is 4.48 ms, while the AscendC bitmask operator reduces the average latency to 0.171 ms.

6493e7626971282da6c32bddbfac3c2a
0b4d7ce4f141cec0bd7e3ac92bff0b9b

Under 16-concurrency tests with the Qwen3-8B model using Eagle3 speculative decoding and structured JSON output:
CPU fallback: TTFT = 157.7 ms, TPOT = 77.69 ms
AscendC bitmask operator: TTFT = 107.5 ms, TPOT = 37.24 ms

CPU Fallback
======================================================================
BENCHMARK RESULTS  (concurrency=16)
======================================================================
  Wall-clock time: 62.75 s

  TTFT (Time To First Token):
    Mean:   157.7 ms
    Median: 139.1 ms
    P95:    262.9 ms
    P99:    266.1 ms
    Std:    45.5 ms
    Min:    130.8 ms
    Max:    266.1 ms

  TPOT (Time Per Output Token):
    Mean:   77.69 ms
    Median: 77.96 ms
    P95:    107.93 ms
    Std:    18.02 ms
    Min:    19.92 ms
    Max:    154.94 ms

  E2E Latency:
    Mean:   5.112 s
    Median: 3.377 s
    P95:    9.149 s
    Std:    8.311 s

  Throughput (wall-clock):
    Output tokens/sec: 151.9
    Requests/sec:      1.59

  Tokens:
    Total prompt tokens:     0
    Total completion tokens: 9530
    Avg completion tokens:   95.3

  Accuracy:
    JSON Parse Success:  95/100 (95.0%)
    Schema Valid:        95/100 (95.0%)

AscendC 
======================================================================
BENCHMARK RESULTS  (concurrency=16)
======================================================================
  Wall-clock time: 40.94 s

  TTFT (Time To First Token):
    Mean:   107.5 ms
    Median: 82.6 ms
    P95:    240.6 ms
    P99:    245.0 ms
    Std:    58.1 ms
    Min:    74.2 ms
    Max:    245.0 ms

  TPOT (Time Per Output Token):
    Mean:   37.24 ms
    Median: 37.31 ms
    P95:    52.20 ms
    Std:    8.49 ms
    Min:    12.50 ms
    Max:    68.99 ms

  E2E Latency:
    Mean:   2.752 s
    Median: 1.661 s
    P95:    4.502 s
    Std:    5.435 s

  Throughput (wall-clock):
    Output tokens/sec: 232.8
    Requests/sec:      2.44

  Tokens:
    Total prompt tokens:     0
    Total completion tokens: 9530
    Avg completion tokens:   95.3

  Accuracy:
    JSON Parse Success:  95/100 (95.0%)
    Schema Valid:        95/100 (95.0%)

welink contact: d00863658

### CODE REVIEW

**Comment by CarterDuan** [comment]
Addressed all review comments. Please review again.

**Comment by iforgetmyname** [comment]
@codex review

**Comment by iforgetmyname** [comment]
> Addressed all review comments. Please review again.

Have you checked the performance after following gemini's reviews?

File: csrc/deepep/ops/op_kernel/dispatch_ffn_combine_kernel/utils/moe_distribute_base.h
Comment (iforgetmyname) [inline csrc/deepep/ops/op_kernel/dispatch_ffn_combine_kernel/utils/moe_distribute_base.h:1]: please revert back unrelated changes

File: csrc/deepep/ops/op_kernel/moe_distribute_base.h
Comment (iforgetmyname) [inline csrc/deepep/ops/op_kernel/moe_distribute_base.h:1]: please revert back unrelated changes

**Comment by CarterDuan** [comment]
> > Addressed all review comments. Please review again.
> 
> Have you checked the performance after following gemini's reviews?

yes, I have done the performance test again

File: csrc/deepep/ops/op_kernel/moe_distribute_base.h
Comment (CarterDuan) [reply csrc/deepep/ops/op_kernel/moe_distribute_base.h:1]: I have revert unrelated changes of two moe files, and done a perfermance test again just now.

==========================================================================================
apply_token_bitmask Performance Report
==========================================================================================
  Warmup: 10 iters, Measurement: 100 iters
  Scenarios: 29
==========================================================================================
Tag                  Description                         NPU(ms)    CPU(ms)    Speedup   NPU BW(GB/s)        Data Volume
------------------------------------------------------------------------------------------------------------------------
prefill_1x128k       Prefill: 1×128256 fp16                1.214    555.600      457.6x           0.44           529.1 KB
prefill_1x128k_bf    Prefill: 1×128256 bf16                1.207    558.387      462.7x           0.44           529.1 KB
prefill_1x128k_f32   Prefill: 1×128256 fp32                1.138    558.177      490.3x           0.92             1.0 MB
prefill_2x128k       Prefill: 2×128256 fp16                1.207   1118.885      926.8x           0.88             1.1 MB
prefill_4x128k       Prefill: 4×128256 fp16                1.205   2242.309     1860.9x           1.76             2.1 MB
prefill_8x128k       Prefill: 8×128256 fp16                1.208   4463.418     3695.5x           3.50             4.2 MB
decode_1x32k         Decode: 1×32000 fp16                  0.341    140.101      410.8x           0.39           132.0 KB
decode_8x32k         Decode: 8×32000 fp16                  0.341   1114.502     3265.6x           3.09             1.1 MB
decode_16x32k        Decode: 16×32000 fp16                 0.342   2230.582     6525.8x           6.18             2.1 MB
decode_32x32k        Decode: 32×32000 fp16                 0.344   4439.781    12891.2x          12.26             4.2 MB
decode_64x32k        Decode: 64×32000 fp16                 7.258   8853.060     1219.8x           1.16             8.4 MB
decode_128x32k       Decode: 128×32000 fp16                3.228  17712.891     5487.1x           5.23            16.9 MB
decode_32x32k_bf     Decode: 32×32000 bf16                 0.342   4425.525    12924.3x          12.34             4.2 MB
decode_32x32k_f32    Decode: 32×32000 fp32                 0.325   4432.240    13622.5x          25.57             8.3 MB
small_1x256          Small: 1×256 fp16                     0.053      1.063       20.0x           0.02             1.1 KB
small_4x256          Small: 4×256 fp16                     0.054      4.443       82.9x           0.08             4.2 KB
small_8x1024         Small: 8×1024 fp16                    0.060     35.311      587.4x           0.56            33.8 KB
small_16x2048        Small: 16×2048 fp16                   0.070    141.962     2019.2x           1.92           135.2 KB
idx_32x32k_8         Indices: 32×32000 idx=8 fp16          0.428   1106.504     2584.8x           9.87             4.2 MB
idx_32x32k_16        Indices: 32×32000 idx=16 fp16         0.447   2223.121     4977.6x           9.46             4.2 MB
idx_1x128k_1         Indices: 1×128256 idx=1 fp16          1.292    556.686      430.8x           0.41           529.1 KB
idx_64x32k_32        Indices: 64×32000 idx=32 fp16         0.503   4447.116     8840.0x          16.79             8.4 MB
idx_128x32k_64       Indices: 128×32000 idx=64 fp16        7.424   8893.822     1198.0x           2.28            16.9 MB
unalign_1x32001      Unalign: 1×32001 fp16                 0.468    138.074      295.0x           0.28           132.0 KB
unalign_16x32001     Unalign: 16×32001 fp16                0.426   2220.116     5208.1x           4.95             2.1 MB
unalign_1x128257     Unalign: 1×128257 fp16                1.339    558.459      417.0x           0.40           529.1 KB
large_batch_512x64   LargeBatch: 512×64 fp16               0.253    141.953      562.0x           0.54           135.2 KB
large_batch_256x128  LargeBatch: 256×128 fp16              0.191    141.547      742.1x           0.71           135.2 KB
large_batch_1024x128 LargeBatch: 1024×128 fp16             0.266    565.409     2129.0x           2.04           540.7 KB

==========================================================================================
Category Summary
==========================================================================================
  Category     Cases    Avg Speedup    Max Speedup   Avg BW(GB/s)
  ------------------------------------------------------------
  decode           8        7043.4x       13622.5x           8.28
  indices          5        3606.3x        8840.0x           7.76
  prefill          6        1315.6x        3695.5x           1.32
  small            4         677.4x        2019.2x           0.65
  stress           3        1144.4x        2129.0x           1.09
  unaligned        3        1973.3x        5208.1x           1.88

  OVERALL         29        3252.9x       13622.5x           4.29
==========================================================================================

File: csrc/apply_token_bitmask/op_kernel/apply_token_bitmask.cpp [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+        this->dtypeSize = dtypeSize;
+
+        uint32_t blockIdx = AscendC::GetBlockIdx();
+        if (blockIdx < blockDim - 1) {
+            this->startRow = blockIdx * rowsPerCore;
+            this->localRows = rowsPerCore;
+        } else {
+            this->startRow = blockIdx * rowsPerCore;
+            this->localRows = rowsLastCore;
+        }
```
Comment (iforgetmyname) [inline csrc/apply_token_bitmask/op_kernel/apply_token_bitmask.cpp:41]: This assignment looks wired when following the calc of `rowsPerCore` and `rowsLastCore`

the calc of two vars are:
```c++
    uint32_t rowsPerCore = static_cast<uint32_t>(numIndices) / blockDim;
    uint32_t rowsLastCore = static_cast<uint32_t>(numIndices) - rowsPerCore * (blockDim - 1);
```

then give the test case of `batch_size=64` and `num_logits=128k`, as `blockDim=40`, we have `rowsPerCore=1` and `rowsLastCore=25` which means for each core you have to process `25` rows!

and in the case of `batch_size=128` you have to process `11` rows per core!

this expains the significant performance drop and why `decode_64x32k` behaves worse than `decode_128x32k`

Comment (CarterDuan) [reply csrc/apply_token_bitmask/op_kernel/apply_token_bitmask.cpp:41]: fix this issue and redo the performance test, get improved on decode_64*32K. Also I runned the pre-commit, please check again. Thanks.

==========================================================================================
apply_token_bitmask Performance Report
==========================================================================================
  Warmup: 10 iters, Measurement: 100 iters
  Scenarios: 29
==========================================================================================
Tag                  Description                         NPU(ms)    CPU(ms)    Speedup   NPU BW(GB/s)        Data Volume
------------------------------------------------------------------------------------------------------------------------
prefill_1x128k       Prefill: 1×128256 fp16                1.211    553.451      456.9x           0.44           529.1 KB
prefill_1x128k_bf    Prefill: 1×128256 bf16                1.211    555.877      459.0x           0.44           529.1 KB
prefill_1x128k_f32   Prefill: 1×128256 fp32                1.150    554.365      482.3x           0.91             1.0 MB
prefill_2x128k       Prefill: 2×128256 fp16                1.212   1111.797      917.5x           0.87             1.1 MB
prefill_4x128k       Prefill: 4×128256 fp16                1.213   2216.667     1826.8x           1.74             2.1 MB
prefill_8x128k       Prefill: 8×128256 fp16                1.213   4436.186     3656.8x           3.49             4.2 MB
decode_1x32k         Decode: 1×32000 fp16                  0.346    138.567      400.1x           0.38           132.0 KB
decode_8x32k         Decode: 8×32000 fp16                  0.346   1101.069     3178.8x           3.05             1.1 MB
decode_16x32k        Decode: 16×32000 fp16                 0.348   2214.292     6358.2x           6.06             2.1 MB
decode_32x32k        Decode: 32×32000 fp16                 0.350   4421.289    12617.1x          12.05             4.2 MB
decode_64x32k        Decode: 64×32000 fp16                 0.640   8847.220    13827.7x          13.20             8.4 MB
decode_128x32k       Decode: 128×32000 fp16                1.216  17700.999    14551.4x          13.89            16.9 MB
decode_32x32k_bf     Decode: 32×32000 bf16                 0.361   4418.982    12240.9x          11.70             4.2 MB
decode_32x32k_f32    Decode: 32×32000 fp32                 0.345   4414.975    12793.4x          24.11             8.3 MB
small_1x256          Small: 1×256 fp16                     0.066      1.147       17.3x           0.02             1.1 KB
small_4x256          Small: 4×256 fp16                     0.066      4.382       66.0x           0.06             4.2 KB
small_8x1024         Small: 8×1024 fp16                    0.075     35.496      471.2x           0.45            33.8 KB
small_16x2048        Small: 16×2048 fp16                   0.085    141.087     1669.2x           1.60           135.2 KB
idx_32x32k_8         Indices: 32×32000 idx=8 fp16          0.446   1103.885     2475.1x           9.47             4.2 MB
idx_32x32k_16        Indices: 32×32000 idx=16 fp16         0.468   2212.817     4726.9x           9.02             4.2 MB
idx_1x128k_1         Indices: 1×128256 idx=1 fp16          1.312    560.065      427.0x           0.40           529.1 KB
idx_64x32k_32        Indices: 64×32000 idx=32 fp16         0.530   4442.996     8375.3x          15.92             8.4 MB
idx_128x32k_64       Indices: 128×32000 idx=64 fp16        0.820   8915.964    10872.8x          20.60            16.9 MB
unalign_1x32001      Unalign: 1×32001 fp16                 0.510    137.930      270.4x           0.26           132.0 KB
unalign_16x32001     Unalign: 16×32001 fp16                0.449   2207.844     4913.2x           4.70             2.1 MB
unalign_1x128257     Unalign: 1×128257 fp16                1.377    557.266      404.6x           0.38           529.1 KB
large_batch_512x64   LargeBatch: 512×64 fp16               0.201    140.923      700.0x           0.67           135.2 KB
large_batch_256x128  LargeBatch: 256×128 fp16              0.201    142.291      708.1x           0.67           135.2 KB
large_batch_1024x128 LargeBatch: 1024×128 fp16             0.228    566.967     2482.7x           2.37           540.7 KB

==========================================================================================
Category Summary
==========================================================================================
  Category     Cases    Avg Speedup    Max Speedup   Avg BW(GB/s)
  ------------------------------------------------------------
  decode           8        9496.0x       14551.4x          10.56
  indices          5        5375.4x       10872.8x          11.09
  prefill          6        1299.9x        3656.8x           1.31
  small            4         555.9x        1669.2x           0.53
  stress           3        1297.0x        2482.7x           1.24
  unaligned        3        1862.7x        4913.2x           1.78

  OVERALL         29        4218.9x       14551.4x           5.48
==========================================================================================

**Review by iforgetmyname — APPROVED** [review]
